crates/circuit/src/vf2.rs

```diff
@@ -1232,22 +1262,82 @@ where
 
     /// Find the unmapped candidates with the highest priority from both graphs.
     ///
-    /// This returns the two candidates and the tracking list of neighbours to the existing mapping
-    /// that they were found in. `None` implies that there were no matching pairs in the lists of
-    /// nodes that are outgoing or incoming neighbors of the map, and therefore we're starting a
-    /// search for a disjoint component.  In order for that to be performant, we'd expect the
-    /// feasiblity checks to have realised that there's edges that will never be satisfied already.
-    fn next_candidates(&mut self) -> Option<(N::NodeId, H::NodeId, Option<OpenList>)> {
-        [Some(OpenList::Out), Some(OpenList::In), None]
-            .into_iter()
-            .filter_map(|list| {
-                Some((
-                    self.needle.first_unmapped(list)?,
-                    self.haystack.first_unmapped(list)?,
-                    list,
-                ))
-            })
-            .next()
+    /// If found, the two candidates are guaranteed to satisfy the most basic compatibility
+    /// requirements with respect to neighboring the partial mapping (i.e. the haystack node will be
+    /// a neighbor in the same direction(s) of the mapping as the needle in an exact or
+    /// induced-subgraph problem, and will be _at least_ a neighbor in the same direction(s) for
+    /// non-induced subgraphs).
+    fn next_candidates(&mut self) -> Option<(N::NodeId, H::NodeId, NeighborKind)> {
+        let mut needle_isolated: Option<N::NodeId> = None;
+        let mut needle_outgoing: Option<N::NodeId> = None;
+        let mut needle_incoming: Option<N::NodeId> = None;
+        let mut needle_candidate_from = |skip: usize| -> Option<(N::NodeId, NeighborKind)> {
```

Comment:
I am a bit confused about the implementation. I believe this closure should return the next potential needle node (in the fixed order we store them), but we want to prioritize nodes already connected (to nodes already in the mapping) over nodes that are not, where we no longer care how the nodes are connected (i.e. incoming, outgoing, both). Does this make sense so far?

Reply:
Yes, that's right.